sdk/python/test_data/pipelines/parallelfor_fan_in/artifacts_complex.py

```diff
@@ -0,0 +1,88 @@
+from typing import List, NamedTuple
+
+from kfp import compiler
+from kfp import dsl
+from kfp.dsl import Dataset
+from kfp.dsl import Input
+from kfp.dsl import Output
+
+
+@dsl.component
+def double(
+    num: int,
+    out_dataset: Output[Dataset],
+):
+    with open(out_dataset.path, 'w') as f:
+        f.write(str(2 * num))
+
+
+@dsl.component
+def add(
+    in_datasets: Input[List[Dataset]],
+    out_dataset: Output[Dataset],
+):
+    nums = []
+    for dataset in in_datasets:
+        with open(dataset.path) as f:
+            nums.append(int(f.read()))
+    with open(out_dataset.path, 'w') as f:
+        f.write(str(sum(nums)))
+
+
+@dsl.component
+def add_two_ints(
+    in_dataset1: Input[Dataset],
+    in_dataset2: Input[Dataset],
+    out_dataset: Output[Dataset],
+):
+    with open(in_dataset1.path) as f:
+        in_dataset1 = int(f.read())
+
+    with open(in_dataset2.path) as f:
+        in_dataset2 = int(f.read())
+
+    with open(out_dataset.path, 'w') as f:
+        f.write(str(in_dataset1 + in_dataset2))
+
+
+@dsl.pipeline
+def add_two_lists_of_datasets(
+    in_datasets1: Input[List[Dataset]],
+    in_datasets2: Input[List[Dataset]],
+) -> Dataset:
+    sum1_task = add(in_datasets=in_datasets1)
+    sum2_task = add(in_datasets=in_datasets2)
+    return add_two_ints(
+        in_dataset1=sum1_task.outputs['out_dataset'],
+        in_dataset2=sum2_task.outputs['out_dataset']).outputs['out_dataset']
+
+
+@dsl.pipeline
+def math_pipeline(
+    threshold: int = 2
+) -> NamedTuple('Outputs', [
+    ('sum', Dataset),
+    ('datasets', List[Dataset]),
+]):
+    with dsl.ParallelFor([1, 2, 3]) as x:
+        double_task1 = double(num=x)
+        with dsl.ParallelFor([1, 2, 3]) as y:
+            with dsl.Condition(y >= threshold):
+                double_task2 = double(num=y)
+    sum_task = add_two_lists_of_datasets(
+        in_datasets1=dsl.Collected(double_task1.outputs['out_dataset']),
+        in_datasets2=dsl.Collected(double_task2.outputs['out_dataset']),
+    )
+    Outputs = NamedTuple('Outputs', [
```

Comment:
nit: Correct me if I'm wrong, I have the impression that while this work, by convention `typing.NamedTuple` is used as a type, while `collections.namedtuple` is used as a constructor.

Reply:
I think this is a typical way to use `typing.NamedTuple`. The `typing` standard library [docs](https://docs.python.org/3/library/typing.html#typing.NamedTuple) describes `typing.NamedTuple` as a "Typed version of collections.namedtuple()". I interpret this as meaning "use this like you would a `collections.namedtuple`". I also see several usage examples that are consistent with this.